src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveRenderedService.cs

```diff
@@ -0,0 +1,127 @@
+using Microsoft.Health.Dicom.Core.Messages.Retrieve;
+using Microsoft.Health.Dicom.Core.Web;
+using Microsoft.IO;
+using SixLabors.ImageSharp;
+
+namespace Microsoft.Health.Dicom.Core.Features.Retrieve;
+public class RetrieveRenderedService : IRetrieveRenderedService
+{
+    private readonly IFileStore _blobDataStore;
+    private readonly IInstanceStore _instanceStore;
+    private readonly IDicomRequestContextAccessor _dicomRequestContextAccessor;
+    private readonly RetrieveConfiguration _retrieveConfiguration;
+    private readonly RecyclableMemoryStreamManager _recyclableMemoryStreamManager;
+    private readonly ILogger<RetrieveRenderedService> _logger;
+
+    public RetrieveRenderedService(
+        IInstanceStore instanceStore,
+        IFileStore blobDataStore,
+        IDicomRequestContextAccessor dicomRequestContextAccessor,
+        IOptionsSnapshot<RetrieveConfiguration> retrieveConfiguration,
+        RecyclableMemoryStreamManager recyclableMemoryStreamManager,
+        ILogger<RetrieveRenderedService> logger)
+    {
+        EnsureArg.IsNotNull(instanceStore, nameof(instanceStore));
+        EnsureArg.IsNotNull(blobDataStore, nameof(blobDataStore));
+        EnsureArg.IsNotNull(dicomRequestContextAccessor, nameof(dicomRequestContextAccessor));
+        EnsureArg.IsNotNull(retrieveConfiguration?.Value, nameof(retrieveConfiguration));
+        EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        _instanceStore = instanceStore;
+        _blobDataStore = blobDataStore;
+        _dicomRequestContextAccessor = dicomRequestContextAccessor;
+        _retrieveConfiguration = retrieveConfiguration?.Value;
+        _recyclableMemoryStreamManager = recyclableMemoryStreamManager;
+        _logger = logger;
+    }
+
+    public async Task<RetrieveRenderedResponse> RetrieveRenderedImageAsync(RetrieveRenderedRequest request, CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+        var partitionKey = _dicomRequestContextAccessor.RequestContext.GetPartitionKey();
+
+        try
+        {
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            AcceptHeader returnHeader = GetValidRenderAcceptHeader(request.AcceptHeaders);
+
+            // this call throws NotFound when zero instance found
+            InstanceMetadata instance = (await _instanceStore.GetInstancesWithProperties(
+                ResourceType.Instance, partitionKey, request.StudyInstanceUid, request.SeriesInstanceUid, request.SopInstanceUid, cancellationToken)).First();
+
+            FileProperties fileProperties = await RetrieveHelpers.CheckFileSize(_blobDataStore, _retrieveConfiguration.MaxDicomFileSize, instance, cancellationToken);
+            Stream stream = await _blobDataStore.GetFileAsync(instance.VersionedInstanceIdentifier, cancellationToken);
```

Comment:
This works for instance, but for frame why are getting the whole. Instead we should get only the specific frame content and then validate whether that frame cotent is more than the maxfilesize we support.

Reply:
May be DicomImage is efficient to hold only the specific frame content in memory. Could you verify that?

Reply:
Leavening as is as we discussed offline and checked fodicom code to confirm they already make optimization there.